packages/peregrine/lib/talons/SignIn/__tests__/useSignIn.spec.js

```diff
@@ -0,0 +1,174 @@
+            {children}
+        </MockedProvider>
+    );
+
+    return renderHook(useSignIn, { wrapper, ...renderHookOptions });
+};
+
+test('returns correct shape', () => {
+    const { result } = renderHookWithProviders();
+
+    expect(result.current).toMatchInlineSnapshot(`
+        Object {
+          "errors": Map {
+            "getUserDetailsQuery" => "getDetails error from redux",
+            "signInMutation" => undefined,
+          },
+          "handleCreateAccount": [Function],
+          "handleForgotPassword": [Function],
+          "handleSubmit": [Function],
+          "isBusy": false,
+          "setFormApi": [Function],
+        }
+    `);
+});
+
+test('handleSubmit triggers waterfall of operations and actions', async () => {
+    const [, { getCartDetails }] = useCartContext();
+    const [, { getUserDetails, setToken }] = useUserContext();
+
+    const { result } = renderHookWithProviders();
+
+    await act(() => result.current.handleSubmit(signInVariables));
+
+    expect(result.current.isBusy).toBe(true);
+    expect(setToken).toHaveBeenCalledWith(authToken);
+    expect(getCartDetails).toHaveBeenCalled();
+    expect(getUserDetails).toHaveBeenCalled();
+});
+
+test('handleSubmit exception is logged and resets state', async () => {
+    const errorMessage = 'Oh no! Something went wrong :(';
+    const [, { getUserDetails, setToken }] = useUserContext();
+    setToken.mockRejectedValue(errorMessage);
+    jest.spyOn(console, 'error');
+
+    const { result } = renderHookWithProviders();
+
+    await act(() => result.current.handleSubmit(signInVariables));
+
+    expect(result.current.isBusy).toBe(false);
+    expect(getUserDetails).not.toHaveBeenCalled();
+    expect(console.error).toHaveBeenCalledWith(errorMessage);
+});
+
+test('handleForgotPassword triggers callbacks', () => {
+    const mockUsername = 'fry@planetexpress.com';
+    const mockApi = {
+        getValue: jest.fn().mockReturnValue(mockUsername)
+    };
+
+    const { result } = renderHookWithProviders();
+    act(() => result.current.setFormApi(mockApi));
+    act(() => result.current.handleForgotPassword());
+
+    expect(initialProps.setDefaultUsername).toHaveBeenCalledWith(mockUsername);
+    expect(initialProps.showForgotPassword).toHaveBeenCalled();
+});
+
+test('handleCreateAccount triggers callbacks', () => {
+    const mockUsername = 'fry@planetexpress.com';
+    const mockApi = {
+        getValue: jest.fn().mockReturnValue(mockUsername)
+    };
+
+    const { result } = renderHookWithProviders();
+    act(() => result.current.setFormApi(mockApi));
+    act(() => result.current.handleCreateAccount());
+
+    expect(initialProps.setDefaultUsername).toHaveBeenCalledWith(mockUsername);
+    expect(initialProps.showCreateAccount).toHaveBeenCalled();
+});
+
+test('mutation error is returned by talon', async () => {
+    const signInErrorMock = {
+        request: signInMock.request,
+        error: new Error('Uh oh! There was an error signing in :(')
```

Comment:
Do we seriously have a sad smiley in the error message 🤣 ?

Reply:
He why not is funny :smile_cat: